### PR TITLE
Adding the recording time when comparing tours with a reference tour

### PR DESCRIPTION
**Pull request by FJBDev**
image

### CODE REVIEW

**Comment by FJBDev** [comment]
I have an issue when using the reference tour and comparing tour perspectives. I keep getting this "A tour is not selected" :
image

I did some research and it seems that the refId of the reference tour is always equal to 0 : 
image


Am I missing something ? Do you have the same issue ?

**Comment by FJBDev** [comment]
> No, not yet but I see you have opened 5 perspectives which can influence the selection provider. I would close all other perspectives to check if this causes the issue.
> 
> You can also debug it to find which selection provides a refId=0
> 

Ok, I will try that.

> Currently I've trouble to build the product. First I replaced Eclipse NL 3.8.2 repository with 4.12 (jar's for 4.13 are not yet available) and now different issues occure, e.g.
> 
> ```
> [ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:1.4.0:compile (default-compile) on project net.tourbook.common: Compilation failure: Compilation failure:
> [ERROR] C:\DAT\mytourbook-BUILD-autocreated\core\net.tourbook.common\src\net\tourbook\common\time\TimeTools.java:[431]
> [ERROR]         return period.toString(DURATION_FORMATTER);
> [ERROR]                       ^^^^^^^^
> [ERROR] Access restriction: The method 'AbstractPeriod.toString(PeriodFormatter)' is not API (restriction on classpath entry 'C:\DAT\mytourbook-BUILD-autocreated\core\net.tourbook.ext.jars\joda-time-2.9.4.jar')
> ```
> 
> or
> 
> ```
> Cannot complete the install because one or more required items could not be found.
>  Software being installed: MyTourbook Product 19.10.0.20191016-1453 (net.tourbook.product 19.10.0.20191016-1453)
>  Missing requirement: toolinggtk.linux.x86_64org.eclipse.equinox.ds 19.10.0.20191016-1453 requires 'bundle org.eclipse.equinox.ds 1.6.0.v20190122-0806' but it could not be found
>  Cannot satisfy dependency:
>   From: MyTourbook Product 19.10.0.20191016-1453 (net.tourbook.product 19.10.0.20191016-1453)
>   To: toolingnet.tourbook.product.configuration [19.10.0.20191016-1453]
>  Cannot satisfy dependency:
>   From: toolingnet.tourbook.product.configuration 19.10.0.20191016-1453
>   To: toolinggtk.linux.x86_64org.eclipse.equinox.ds [19.10.0.20191016-1453]
> ```
> 
> Depending which parameters I change, e.g. tycho from 0.25 to 1.4 and so on.
> So it can take a while until I can fix these issues.

Weird, didn't you tell me that the build was working [25 days ago](https://github.com/wolfgang-ch/mytourbook/pull/110/files#r326859540) ?

**Comment by wolfgang-ch** [comment]
> Weird, didn't you tell me that the build was working [25 days ago](https://github.com/wolfgang-ch/mytourbook/pull/110/files#r326859540) ?

It worked

**Comment by wolfgang-ch** [comment]
> Since I haven't been using that feature for a long time maybe I am missing something ? Do you have the same issue ?

Now I saw this issue when starting with an empty tour database and fixed it in https://github.com/wolfgang-ch/mytourbook/commit/489c321dd8e1d82a1d7357803743dacc8f60c906

When I reviewed the code in TourDatabase.java then I saw that the new recording time field is not set for an empty database -> so I tested an empty database and when a compared tour is saved then this issue occures

```07:31:48,217  WARN JDBCExceptionReporter:100 - SQL Error: 20000, SQLState: 42X14
07:31:48,217 ERROR JDBCExceptionReporter:101 - 'TOURRECORDINGTIME' is not a column in table or VTI 'USER.TOURCOMPARED'.
javax.persistence.PersistenceException: org.hibernate.exception.SQLGrammarException: could not insert: [net.tourbook.data.TourCompared]
	at org.hibernate.ejb.AbstractEntityManagerImpl.throwPersistenceException(AbstractEntityManagerImpl.java:614)
	at org.hibernate.ejb.AbstractEntityManagerImpl.persist(AbstractEntityManagerImpl.java:226)
	at net.tourbook.ui.views.tourCatalog.TourCompareManager.saveComparedTourItem(TourCompareManager.java:206)
	at net.tourbook.ui.views.tourCatalog.TourCompareResultView.saveCompareResults(TourCompareResultView.java:1618)
	at net.tourbook.ui.views.tourCatalog.ActionSaveComparedTours.run(ActionSaveComparedTours.java:40)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:412)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4173)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1057)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:3986)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3585)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1160)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1049)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:633)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:557)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
	at net.tourbook.application.Application.start(Application.java:36)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:400)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:660)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:597)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1468)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1441)
```

File: bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java [hunk 1]
```diff
@@ -7317,6 +7317,16 @@ private int updateDbDesign_039_to_040(final Connection conn, final SplashManager
...
 
+// SET_FORMATTING_ON
+         }
+
+         if (isColumnAvailable(conn, TABLE_TOUR_COMPARED, "tourRecordingTime") == false) { //$NON-NLS-1$
+
+// SET_FORMATTING_OFF
+
+            // Add new columns
+            SQL.AddCol_Float(stmt, TABLE_TOUR_COMPARED, "tourRecordingTime", DEFAULT_0); //$NON-NLS-1$
```
Comment (wolfgang-ch) [inline bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java]: This field is an integer value and not a float

File: bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java [hunk 2]
```diff
@@ -7317,6 +7317,16 @@ private int updateDbDesign_039_to_040(final Connection conn, final SplashManager
...
+
```
Comment (wolfgang-ch) [inline bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java]: There is already one column checked for this version, so I think it is not necessary to check a second column

**Comment by wolfgang-ch** [comment]
For existing compared tours, the new `tourRecordingTime` values are 0. Shouldn't this be updated in a `updateDbDesign_039_to_040_PostUpdate` method to the correct value ?

File: bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java (hunk 1 above)
Comment (wolfgang-ch) [reply bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java]: For version 19.10, I can only send an e-mail to the translators to do the translation, when the tour database is final.

Comment (FJBDev) [reply bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java]: I will correct

File: bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java (hunk 2 above)
Comment (FJBDev) [reply bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java]: Right, I will remove

**Comment by FJBDev** [comment]
> 
> 
> For existing compared tours, the new `tourRecordingTime` values are 0. Shouldn't this be updated in a `updateDbDesign_039_to_040_PostUpdate` method to the correct value ?

I didn't think about that, I will do it

**Comment by FJBDev** [comment]
> 
> 
> > Since I haven't been using that feature for a long time maybe I am missing something ? Do you have the same issue ?
> 
> Now I saw this issue when starting with an empty tour database and fixed it in [489c321](https://github.com/wolfgang-ch/mytourbook/commit/489c321dd8e1d82a1d7357803743dacc8f60c906)
> 
> When I reviewed the code in TourDatabase.java then I saw that the new recording time field is not set for an empty database -> so I tested an empty database and when a compared tour is saved then this issue occures
> 
> ```
> 07:31:48,217 ERROR JDBCExceptionReporter:101 - 'TOURRECORDINGTIME' is not a column in table or VTI 'USER.TOURCOMPARED'.
> javax.persistence.PersistenceException: org.hibernate.exception.SQLGrammarException: could not insert: [net.tourbook.data.TourCompared]
> 	at org.hibernate.ejb.AbstractEntityManagerImpl.throwPersistenceException(AbstractEntityManagerImpl.java:614)
> 	at org.hibernate.ejb.AbstractEntityManagerImpl.persist(AbstractEntityManagerImpl.java:226)
> 	at net.tourbook.ui.views.tourCatalog.TourCompareManager.saveComparedTourItem(TourCompareManager.java:206)
> 	at net.tourbook.ui.views.tourCatalog.TourCompareResultView.saveCompareResults(TourCompareResultView.java:1618)
> 	at net.tourbook.ui.views.tourCatalog.ActionSaveComparedTours.run(ActionSaveComparedTours.java:40)
> 	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
> 	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
> 	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:412)
> 	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
> 	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4173)
> 	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1057)
> 	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:3986)
> 	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3585)
> 	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1160)
> 	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
> 	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1049)
> 	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
> 	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:633)
> 	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
> 	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:557)
> 	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
> 	at net.tourbook.application.Application.start(Application.java:36)
> 	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
> 	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
> 	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
> 	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:400)
> 	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:660)
> 	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:597)
> 	at org.eclipse.equinox.launcher.Main.run(Main.java:1468)
> 	at org.eclipse.equinox.launcher.Main.main(Main.java:1441)
> ```

Hmmm, I tried with a non empty and empty DB and I didn't see any issues, I will try again.

File: bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java (hunk 2 above)
Comment (FJBDev) [reply bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java]: Fixed, see my [PR](https://github.com/wolfgang-ch/mytourbook/pull/153/)

File: bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java (hunk 1 above)
Comment (FJBDev) [reply bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java]: Fixed, see my [PR](https://github.com/wolfgang-ch/mytourbook/pull/153/)

**Comment by FJBDev** [comment]
I fixed all those comments. See my [PR](https://github.com/wolfgang-ch/mytourbook/pull/153/)

**Pull request merged**